### DISCUSSION TITLE
gRPC Using Windows and RPi

### DISCUSSION THREAD

**Discussion by myvikes** [mattjlewis/diozero#195] [General]
I'm using the gRPC explained [here](https://www.diozero.com/concepts/remotecontrol.html) and all works fine except that I'm trying to figure out how when I restart my app on my Windows PC using Eclipse, I can do so such that I don't have to also first restart the GrpcServer on my RPi also.  If I don't restart the server, I receive errors such as 

> com.diozero.api.RuntimeIOException: Error in I2C open: I2C device already provisioned

Any suggestions or direction would be greatly appreciated.

Thanks!

**Comment by myvikes** [comment]
NM...i got this to work...it seems a simple System.exit(0); automagically closes all the diozero devices. :-)

**Comment by mattjlewis** [comment]
I've done my best to auto-shutdown things but you're right - System.exit(0) is always a good idea if the app doesn't exit on its own after closing all devices. I strongly recommend using try with resources whenever you can - I turn on compiler warnings (in Eclipse) so that I can see if there is a resource that isn't being closed. Closing the default device factory should cleanup most of the diozero stuff, however, there can be other non-daemon threads (such as gRPC or Netty) that block JVM shutdown.
